### PR TITLE
Null check before accessing current document

### PR DESCRIPTION
**Pull request by datho7561**
Fix "Cannot read property document of undefined" error during launch.

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: src/variableSubstitution.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ const VARIABLE_SUBSTITUTIONS: VariableSubstitution[] = [
 export function getVariableSubstitutedAssociations(associations: XMLFileAssociation[]): XMLFileAssociation[] {
 
   // Collect properties needed to resolve variables
-  const currentFile: TextDocument = window.activeTextEditor.document;
+  const currentFile: TextDocument = !!window.activeTextEditor ? window.activeTextEditor.document: undefined;
```
Comment (angelozerr) [inline src/variableSubstitution.ts:102]: What about writing this code ?

```ts
const currentFile: TextDocument = window.activeTextEditor && window.activeTextEditor.document;
```

Comment (angelozerr) [reply src/variableSubstitution.ts:102]: I think we should check the languageId too:

`window.activeTextEditor.document.languageId`

Comment (fbricon) [reply src/variableSubstitution.ts:102]: have you tried window?.activeTextEditor.document

Comment (fbricon) [reply src/variableSubstitution.ts:102]: see https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

File: src/variableSubstitution.ts [hunk 2]
```diff
@@ -99,7 +99,7 @@ const VARIABLE_SUBSTITUTIONS: VariableSubstitution[] = [
...
   const currentFileUri: string = currentFile && currentFile.uri.fsPath;
   const currentWorkspace: WorkspaceFolder = workspace.getWorkspaceFolder(currentFile && currentFile.uri);
```
Comment (angelozerr) [inline src/variableSubstitution.ts:104]: It seems that `workspace.getWorkspaceFolder(null) `throw an error.

File: src/variableSubstitution.ts (hunk 1 above)
Comment (datho7561) [reply src/variableSubstitution.ts:102]: @Fred Doesn't work, I get a runtime error, but I'll check the langaugeId

Comment (angelozerr) [reply src/variableSubstitution.ts]: @datho7561 here the fix which is working for me:

```ts
const currentFile: TextDocument = window.activeTextEditor && window.activeTextEditor.document;
const currentFileUri: string = currentFile && currentFile.uri.fsPath;
  const currentWorkspace: WorkspaceFolder = currentFile && currentFile.uri && workspace.getWorkspaceFolder(currentFile.uri);
```

File: src/extension.ts [hunk 3]
```diff
@@ -429,6 +429,8 @@ export function activate(context: ExtensionContext) {
     xml['xml']['format']['trimTrailingWhitespace'] = workspace.getConfiguration('files').get('trimTrailingWhitespace', false);
     xml['xml']['format']['insertFinalNewline'] = workspace.getConfiguration('files').get('insertFinalNewline', false);
 
+    // console.log(xml['xml']['symbols']['maxItems'])
```
Comment (angelozerr) [inline src/extension.ts:432]: remove this comment

File: src/variableSubstitution.ts [hunk 4]
```diff
@@ -99,9 +99,12 @@ const VARIABLE_SUBSTITUTIONS: VariableSubstitution[] = [
 export function getVariableSubstitutedAssociations(associations: XMLFileAssociation[]): XMLFileAssociation[] {
 
   // Collect properties needed to resolve variables
-  const currentFile: TextDocument = window.activeTextEditor.document;
-  const currentFileUri: string = currentFile && currentFile.uri.fsPath;
-  const currentWorkspace: WorkspaceFolder = workspace.getWorkspaceFolder(currentFile && currentFile.uri);
+  const currentFile: TextDocument = (!!window.activeTextEditor && !!window.activeTextEditor.document && window.activeTextEditor.document.languageId === 'xml') ? window.activeTextEditor.document : undefined;
+  const currentFileUri: string = currentFile.uri && currentFile.uri.fsPath;
```
Comment (angelozerr) [inline src/variableSubstitution.ts:103]: I think you can write in one line:

const currentWorkspace: WorkspaceFolder = currentFile && currentFile.uri && workspace.getWorkspaceFolder(currentFile.uri);

**Comment by angelozerr** [comment]
> have you tried window?.activeTextEditor.document

@fbricon  I tried it but it doesn't work?

File: src/variableSubstitution.ts [hunk 5]
```diff
@@ -99,9 +99,9 @@ const VARIABLE_SUBSTITUTIONS: VariableSubstitution[] = [
 export function getVariableSubstitutedAssociations(associations: XMLFileAssociation[]): XMLFileAssociation[] {
 
   // Collect properties needed to resolve variables
-  const currentFile: TextDocument = window.activeTextEditor.document;
-  const currentFileUri: string = currentFile && currentFile.uri.fsPath;
-  const currentWorkspace: WorkspaceFolder = workspace.getWorkspaceFolder(currentFile && currentFile.uri);
+  const currentFile: TextDocument = window.activeTextEditor && window.activeTextEditor.document && window.activeTextEditor.document.languageId === 'xml' && window.activeTextEditor.document;
+  const currentFileUri: string = currentFile.uri && currentFile.uri.fsPath;
```
Comment (angelozerr) [inline src/variableSubstitution.ts:103]: It should be:

`const currentFileUri: string = currentFile && currentFile.uri && currentFile.uri.fsPath;`

File: src/variableSubstitution.ts [hunk 6]
```diff
@@ -99,9 +99,9 @@ const VARIABLE_SUBSTITUTIONS: VariableSubstitution[] = [
 export function getVariableSubstitutedAssociations(associations: XMLFileAssociation[]): XMLFileAssociation[] {
 
   // Collect properties needed to resolve variables
-  const currentFile: TextDocument = window.activeTextEditor.document;
-  const currentFileUri: string = currentFile && currentFile.uri.fsPath;
-  const currentWorkspace: WorkspaceFolder = workspace.getWorkspaceFolder(currentFile && currentFile.uri);
+  const currentFile: TextDocument = window.activeTextEditor && window.activeTextEditor.document && window.activeTextEditor.document.languageId === 'xml' && window.activeTextEditor.document;
```
Comment (angelozerr) [inline src/variableSubstitution.ts:102]: Sorry this code doesn't work,it currentFile is equals to false when the condition is not respected although it should return undefined.

**Comment by angelozerr** [comment]
@datho7561 please see my draft PR which fix all problems https://github.com/redhat-developer/vscode-xml/pull/368

**Comment by datho7561** [comment]
Superseded by #368

**Pull request closed without merging**